server/src/features/DocumentSymbolRequestHandler.ts

```diff
@@ -0,0 +1,105 @@
+import { CompletionItemKind, DocumentSymbol, DocumentSymbolParams, SymbolKind } from 'vscode-languageserver';
+
+import { YmlDefinitionProvider } from '../definitions';
+import { AbstractYmlObject } from '../yml-objects';
+
+export function documentSymbolRequestHandler(definitionsProvider: YmlDefinitionProvider) {
+    return (_params: DocumentSymbolParams) => {
+        return buildDocumentSymbolsList(_params.textDocument.uri, definitionsProvider);
+    };
+}
+
+/**
+ * Build the list of all available YML symbols that are in a file.
+ *
+ * @param uri the file URI
+ *
+ * @returns the list of available YML symbols.
+ */
+export function buildDocumentSymbolsList(uri: string, definitionsProvider: YmlDefinitionProvider): DocumentSymbol[] {
+    // Get the YML objects definitions (for which we have a location) that are in this file.
+    const filter = (elem: AbstractYmlObject) => elem.uri === uri && !!elem.definitionLocation;
+    let currentDocSymbols = definitionsProvider.definitions.filter(filter);
+    // Also get the YML objects implementations (with a location too) that are in this file.
+    currentDocSymbols = currentDocSymbols.concat(definitionsProvider.implementations.filter(filter));
+
+    const parentToChildren = new Map<AbstractYmlObject, AbstractYmlObject[]>();
+    for (const elem of currentDocSymbols) {
+        if (!elem.sourceElement) {
+            // Current element has no parent.
+            if (!parentToChildren.has(elem)) {
+                // Current element has no parent and this is the first time we encounter it.
+                parentToChildren.set(elem, []);
+            } else {
+                // Current element is a an already known parent element. No need to add it again.
```

Comment:
* typo (a an)
* don’t get it. You switch from “it has a parent” to “it is a parent” :o

Reply:
I'll rework this comment, then.

Reply:
Done.